### PR TITLE
Updated list of Pokemon in naturalEnemies

### PR DESCRIPTION
**Pull request by pkmnsnfrn**
## Description

# Current Problem
In the current version of expansion, the list of Pokémon in [`naturalEnemies`](https://github.com/rh-hideout/pokeemerald-expansion/blob/e2811035752b8db36ad8a85be35b72260b5c63b8/src/battle_controller_opponent.c#L597) is very small. It is populated during the actual data from SoS / Horde Battles.

# Proposed Solution
The Pokédex in the core series game details more relationships between Pokémon species. Given that fangame developers are likely to have more diverse wild battles than Game Freak, the Pokédex entries should be reflected in the gameplay. [Bulbapedia has cataloged this list of the relationships between prey and predators.](https://bulbapedia.bulbagarden.net/wiki/Pok%C3%A9mon_food#Pok.C3.A9mon_eating_other_Pok.C3.A9mon)

`naturalEnemies` is updated to use this list. This is all gated behind a config, `B_WILD_NATURAL_ENEMIES_POKEDEX`.

~~## Changes From Original~~
~~According to the Pokédex, Mareanie eats Corsola.~~
~~In SOS Battles, Mareanie will attack allied Corsola, but Corsola will not attack allied Mareanie.~~
~~In expansion, this was interpreted as "Corsola should not attack Mareanie".~~

~~According to the Pokédex, Heatmor eats Durant.~~
~~In Horde Battles, allied Heatmor and Durant target each other.~~
~~In expansion, this was interpreted as "Heatmor and Durant should attack each other".~~

~~Horde Battles displays there is precedent for prey fighting back, so this has been reflected in the code by removing one-directional relationships, and allowing both sides of the conflict to attack each other.~~

# Testing

[.sav file](https://cdn.discordapp.com/attachments/960765412268716042/1180390815097425980/pokeemerald.sav?ex=657d3f9c&is=656aca9c&hm=445d51d29fb550aa1d8213b7055f9c7087958615200620adc141bd62e5d7bd44&)

## data/scripts/debug.inc
 ```diff
Debug_EventScript_Script_1::
+    givemon SPECIES_STARMIE, 100, 0
+    givemon SPECIES_EXEGGUTOR, 100, 0
+    givemon SPECIES_SNORLAX, 100, 0
+    givemon SPECIES_TAUROS, 100, 0
+    call Debug_CheatStart
    end

Debug_EventScript_Script_2::
// Zangoose and Seviper will always attack each other.
+    special HealPlayerParty
+    setwildbattle SPECIES_SEVIPER, 10, 0, SPECIES_ZANGOOSE, 10, 0
+    dowildbattle
    end

Debug_EventScript_Script_3::
// B_WILD_NATURAL_ENEMIES_POKEDEX == TRUE, Gumshoos and Rattata will attack each other.
// B_WILD_NATURAL_ENEMIES_POKEDEX == FALSE, Gumshoos and Rattata will NOT attack each other.
+    special HealPlayerParty
+    setwildbattle SPECIES_GUMSHOOS, 27, 0, SPECIES_RATTATA, 27, 0
+    dowildbattle
    end

Debug_EventScript_Script_4::
// Mareanie will always attack Corsola. Corsola will never attack Mareanie.
+    special HealPlayerParty
+    setwildbattle SPECIES_MAREANIE, 10, 0, SPECIES_CORSOLA, 90, 0
+    dowildbattle
    end

Debug_EventScript_Script_5::
// Pikipek and Flygon will never attack each other.
+    special HealPlayerParty
+    setwildbattle SPECIES_JOLTEON, 40, 0, SPECIES_FLAREON, 40, 0
+    dowildbattle
    end
```

## Showcase
In each video, the player uses the debug scripts to start a wild battle for each of the scenarios listed above.
* [`B_WILD_NATURAL_ENEMIES_POKEDEX`  == `TRUE`](https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/9ea31805-e0dc-4b7d-bb87-33eefe19e296)
* [`B_WILD_NATURAL_ENEMIES_POKEDEX`  == `FALSE`](https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/83df02f0-26b0-4992-b5b8-bfe290a3a399)

## **Discord contact info**
pkmnsnfrn

### CODE REVIEW

**Comment by AsparagusEduardo** [comment]
So, are you changing it so that Corsola attacks Mareanie? That's not official behavior.

Similarly, we never see the new prey/predator relationships in proper gameplay besides the current ones, so I wouldn't have these enabled by default.

File: src/battle_controller_opponent.c [hunk 1]
```diff
@@ -637,6 +617,184 @@ static void OpponentHandleChooseMove(u32 battler)
     }
 }
 
+static const u16 naturalEnemies[][3] =
+{
+    // Attacker              // Target
+    {SPECIES_ZANGOOSE,       SPECIES_SEVIPER},
+    {SPECIES_HEATMOR,        SPECIES_DURANT},
+    {SPECIES_SABLEYE,        SPECIES_CARBINK},
+    {SPECIES_MAREANIE,       SPECIES_CORSOLA},
```
Comment (AlexOn1ine) [inline src/battle_controller_opponent.c:626]: I don't think this issue is specific to this PR but with the species refactor, if any of those species are disabled the code wont compile, I believe.

Comment (AsparagusEduardo) [reply src/battle_controller_opponent.c:626]: Nah, that shouldn't be an issue. The species refactor doesn't remove the species defines.

File: src/battle_controller_opponent.c [hunk 2]
```diff
@@ -637,6 +617,184 @@ static void OpponentHandleChooseMove(u32 battler)
...
+{
+    u32 i;
+    u16 speciesAttacker = gBattleMons[battler].species;
+    u16 speciesTarget = gBattleMons[GetBattlerAtPosition(BATTLE_PARTNER(battler))].species;
+
+    if ((speciesAttacker == SPECIES_CARBINK) && (speciesTarget == SPECIES_SABLEYE))
+        return FALSE;
+
+    if ((speciesAttacker == SPECIES_CORSOLA) && (speciesTarget == SPECIES_MAREANIE))
+        return FALSE;
```
Comment (AsparagusEduardo) [inline src/battle_controller_opponent.c:777]: These 2 fall into the "predator vs prey" category, but all other mon that you classified as such don't behave this way.
I would recommend having separate arrays for the different types of interactions.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/battle_controller_opponent.c (hunk 2 above)
Comment (pkmnsnfrn) [reply src/battle_controller_opponent.c]: https://github.com/rh-hideout/pokeemerald-expansion/pull/3622/commits/1144b9c250f8f2f2e7ee19e3e5333c3647dedd07

File: src/battle_controller_opponent.c [hunk 3]
```diff
@@ -617,13 +618,16 @@ static void OpponentHandleChooseMove(u32 battler)
     }
 }
 
+static const u16 predatorPrey[][3] =
+{
+    {SPECIES_SABLEYE,        SPECIES_CARBINK},
+    {SPECIES_MAREANIE,       SPECIES_CORSOLA},
+};
+
 static const u16 naturalEnemies[][3] =
```
Comment (AsparagusEduardo) [inline src/battle_controller_opponent.c]:
```suggestion
static const u16 predatorPrey[][2] =
{   // Predator              Prey
    {SPECIES_SABLEYE,        SPECIES_CARBINK},
    {SPECIES_MAREANIE,       SPECIES_CORSOLA},
};

static const u16 naturalEnemies[][2] =
```

File: src/battle_controller_opponent.c [hunk 4]
```diff
@@ -637,6 +618,190 @@ static void OpponentHandleChooseMove(u32 battler)
...
+    {SPECIES_ELEKID,         SPECIES_TOGEDEMARU},
+    {SPECIES_DWEBBLE,        SPECIES_ROGGENROLA},
+    {SPECIES_DWEBBLE,        SPECIES_ROLYCOLY},
+    {SPECIES_JOLTIK,         SPECIES_YAMPER},
+    {SPECIES_CUTIEFLY,       SPECIES_GOSSIFLEUR},
+    {SPECIES_TOGEDEMARU,     SPECIES_ELEKID},
+    {SPECIES_LECHONK,        SPECIES_APPLETUN},
+    {SPECIES_TINKATUFF,      SPECIES_PAWNIARD},
+    {SPECIES_TINKATUFF,      SPECIES_BISHARP},
+    {SPECIES_TINKATON,       SPECIES_CORVIKNIGHT},
```
Comment (AsparagusEduardo) [inline src/battle_controller_opponent.c:767]: Shouldn't these be moved to the `predatorPrey` array?

File: src/battle_controller_opponent.c [hunk 5]
```diff
@@ -637,6 +618,190 @@ static void OpponentHandleChooseMove(u32 battler)
...
+    {SPECIES_CORVIKNIGHT,    SPECIES_BUNNELBY},
+    {SPECIES_APPLETUN,       SPECIES_CUTIEFLY},
+    {SPECIES_CRAMORANT,      SPECIES_ARROKUDA},
+    {SPECIES_CENTISKORCH,    SPECIES_SCOLIPEDE},
+    {SPECIES_WATTREL,        SPECIES_ARROKUDA},
+    {SPECIES_VELUZA,         SPECIES_WIGLETT},
+    {SPECIES_GRAFAIAI,       SPECIES_SCATTERBUG},
+    // Rivals
+    {SPECIES_CUTIEFLY,       SPECIES_BUTTERFREE},
+    {SPECIES_SHIINOTIC,      SPECIES_PARASECT},
```
Comment (AsparagusEduardo) [inline src/battle_controller_opponent.c:735]: What's the difference between these and the ones above?

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
Can you please link to your sources for these interactions?
I really don't want to look up every single dex entries for all these species.

**Comment by GraionDilach** [comment]
> Can you please link to your sources for these interactions?

https://bulbapedia.bulbagarden.net/wiki/Pok%C3%A9mon_food#Known_predatory_relation

**Pull request closed without merging**